wrappers/python/pyrealsense2.cpp

```diff
@@ -54,25 +54,73 @@ PYBIND11_MODULE(NAME, m) {
         .def("__str__", &rs2::log_message::raw)
         .def("__repr__", &rs2::log_message::full);
 
-    m.def("log_to_callback",
-        [](rs2_log_severity min_severity, std::function<void(rs2_log_severity, rs2::log_message)> callback)
+    // We want to enable Python callbacks for logging, but need to be careful:
+    // The machanism used by librealsense keeps a pointer to an object that is then released
+    // on destruction/exit. Usually this works fine, except that here, with Python and its GIL,
+    // Pybind tries to acquire the GIL when the thread state is no longer valid and we get
+    // into an infinite wait.
+#if 0
+    m.def( "log_to_callback",
+           []( rs2_log_severity min_severity, std::function< void( rs2_log_severity, rs2::log_message ) > callback )
+           {
+               py::gil_scoped_release gil;
+               rs2::log_to_callback( min_severity,
+                                     [callback]( rs2_log_severity severity, rs2::log_message const & msg ) noexcept
+                                     {
+                                         py::gil_scoped_acquire gil;
+                                         callback( severity, msg );
+                                     } );
+           } );
+#else
+    // Instead, as a workaround, we override the usual mechanism to intentionally not free up
+    // if we see the Python thread state isn't valid (see release() below):
+    class py_log_callback : public rs2::log_callback
+    {
+        typedef rs2::log_callback super;
+
+    public:
+        py_log_callback( log_fn && on_log )
+            : super( std::move( on_log ) )
+        {
+        }
+
+        void on_log( rs2_log_severity severity, rs2_log_message const & msg ) noexcept override
         {
-            rs2::log_to_callback( min_severity,
-                [callback]( rs2_log_severity severity, rs2::log_message const & msg ) noexcept
-                {
-                    try
-                    {
-                        // We're not being called from Python but instead are calling it,
-                        // we need to acquire it to not have issues with other threads...
-                        py::gil_scoped_acquire gil;
-                        callback( severity, msg );
```

Comment:
We used to call the `callback` func which I am not sure where we got it from and now we call `on_log`.
Can you explain?

Reply:
`callback` was part of the lambda, passed in from outside when `log_to_callback` was called.
`on_log` is called per message log. The callback is stored in the `on_log_function` inside `log_callback`, and invoked from the base class's `on_log`, which is why we call it.